Guidelines.md

```diff
@@ -0,0 +1,319 @@
+-   Strings (valid Unicode only)
+-   Finite Float64 numbers (no Infinity, NaN, -0, or BigInt)
+    \*Booleans
+-   Implicit undefined (optional properties elided by JSON serialization)
+-   Null (as empty root or placeholder in arrays – see next guideline.)
+
+This subset of JavaScript types forms a minimal but complete data model that is well supported across data stores and transport protocols.
+
+[GraphQL](https://graphql.org/) is a good example of a framework that embraces JSON as the “lowest common denominator”.
+It's worth studying if you’re interested in applying this principle in a cross-language environment.
+
+#### ✔ PREFER `undefined` over `null`
+
+`undefined` is the preferred type for uninitialized variables, missing keys, or a sentinel representing an empty state.
+Optional properties and arguments, which are implicitly `undefined`, are generally preferred over explicit `undefined`.
+
+There are, however, a few cases where `null` continues to be appropriate.
+The primary use case is JSON serializable data where "implicit undefined" is not an option.
+These are:
+
+-   The root of an empty tree ("null”)
+-   An empty placeholder in arrays ("[null, 3]")
+
+#### ✔ AVOID distinguishing between implicit and explicit undefined values
+
+In the JavaScript language it is possible to distinguish between a non-existing property and a property that has been explicitly set to the `undefined` value.
+You should interpret these identically when reading.
+
+Do not go out of your way elide explicit `undefine`s, except in cases where it improves efficiency (such as serialization).
+Otherwise, you should follow the natural behavior of `Object.keys()`, `foo = bar`, etc. which will preserve the explicit undefined.
+
+#### ✔ DO align copy/iteration behavior with Object.keys()
+
+The prevailing convention is to use `Object.keys()` when iterating or copying data, which includes only string-keyed [enumerable properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) that are owned by the object (not inherited).
+
+Note that it is expected that private metadata attached via `symbol` keys will be elided when iterated, copied, transmitted, or persisted.
+
+### Errors
+
+#### ✔ DO limit runtime error checking to non-obvious/hard to diagnose errors
+
+Runtime error checking is reserved for non-obvious errors that are difficult to diagnose without runtime assistance.
+You should assume that developers have common sense and do not exploit quirks or intentionally circumvent the type system.
+
+#### ✘ DO NOT use assertions for validating user input
+
+An assertion failure indicates a bug in the Fluid Framework itself, not the user's code.
+User errors should by signaled by throwing an instance of `Error`, `TypeError`, `ReferenceError`, `RangeError`, `AggregateError` or an appropriate subclass (other built-in error types are reserved for language parsing errors).
```

Comment:
we have a usage error class it could be useful to call out here

Reply:
I pointed to the module with the error definitions in core-interfaces as an example, which includes `UsageError` and a few others.